src/core/Base.ts

```diff
@@ -91,43 +116,58 @@ export class Base {
     this.items.wiki = JSON.parse(fs.readFileSync(join(__dirname, "assets", "items_info_new.json"), "utf-8")) as ItemsInfo[];
   }
 
-  public async saveAll(disconnectAll = false) {
+  public async saveAll(disconnectAll = false): Promise<boolean> {
     consola.info(`Saving ${this.cache.peers.size} peers & ${this.cache.worlds.size} worlds`);
 
-    this.saveWorlds();
-    this.savePlayers(disconnectAll);
+    const worldsSaved = await this.saveWorlds();
+    const playersSaved = await this.savePlayers(disconnectAll);
+
+    return worldsSaved && playersSaved;
   }
 
   public async saveWorlds() {
-    if (this.cache.worlds.size === 0) process.exit();
-    else {
-      let o = 0;
-      this.cache.worlds.forEach(async (wrld) => {
-        const world = new World(this, wrld.name);
-        if (typeof world.worldName === "string") await world.saveToDatabase();
-        else consola.warn(`Oh no there's undefined (${o}) world, skipping..`);
-
-        o += 1;
-        if (o === this.cache.worlds.size) process.exit();
-      });
+    try {
+      let savedCount = 0;
+      for (const [name, world] of this.cache.worlds) {
+        const wrld = new World(this, world.name);
+        if (typeof wrld.worldName === "string") await wrld.saveToDatabase().catch((e) => consola.error(e));
+        else consola.warn(`Oh no there's undefined (${savedCount}) world, skipping..`);
+        savedCount++;
+      }
+      consola.success(`Saved ${savedCount} worlds`);
+      return true;
+    } catch (err) {
+      consola.error(`Failed to save worlds: ${err}`);
+      return false;
     }
   }
 
   public async savePlayers(disconenctAll: boolean) {
-    if (this.cache.peers.size === 0) process.exit();
-    else {
-      let i = 0;
-      this.cache.peers.forEach(async (p) => {
-        const player = new Peer(this, p.netID);
+    try {
+      let savedCount = 0;
+      for (const [_, peer] of this.cache.peers) {
+        const player = new Peer(this, peer.netID);
         await player.saveToDatabase();
         if (disconenctAll) {
           player.disconnect("now");
         } else {
-          // send onconsolemessage for auto saving
+          player.every((p) => {
```

Comment:
Why this there's a nested loop here?

Reply:
i see my bad, didnt notice if that's on a loop